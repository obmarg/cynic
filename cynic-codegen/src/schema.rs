use crate::{FieldArgument, TypeIndex};

// Alias all the graphql_parser schema types so we don't have to specify generic parameters
// everywhere
pub type Document = graphql_parser::schema::Document<'static, String>;
pub type Type = graphql_parser::schema::Type<'static, String>;
pub type Field = graphql_parser::schema::Field<'static, String>;
pub type Definition = graphql_parser::schema::Definition<'static, String>;
pub type TypeDefinition = graphql_parser::schema::TypeDefinition<'static, String>;
pub type ObjectType = graphql_parser::schema::ObjectType<'static, String>;
pub type EnumType = graphql_parser::schema::EnumType<'static, String>;
pub type InterfaceType = graphql_parser::schema::InterfaceType<'static, String>;
pub type InputObjectType = graphql_parser::schema::InputObjectType<'static, String>;
pub type ScalarType = graphql_parser::schema::ScalarType<'static, String>;
pub type UnionType = graphql_parser::schema::UnionType<'static, String>;
pub type InputValue = graphql_parser::schema::InputValue<'static, String>;
pub type EnumValue = graphql_parser::schema::EnumValue<'static, String>;

/// Loads a schema from a filename, relative to CARGO_MANIFEST_DIR if it's set.
pub fn load_schema(filename: impl AsRef<std::path::Path>) -> Result<Document, SchemaLoadError> {
    use std::path::PathBuf;
    let mut pathbuf = PathBuf::new();

    if let Ok(manifest_dir) = std::env::var("CARGO_MANIFEST_DIR") {
        pathbuf.push(manifest_dir);
    } else {
        pathbuf.push(std::env::current_dir()?);
    }
    pathbuf.push(filename);

    let schema = std::fs::read_to_string(&pathbuf)
        .map_err(|_| SchemaLoadError::FileNotFound(pathbuf.to_str().unwrap().to_string()))?;

    parse_schema(&schema)
}

pub(crate) fn parse_schema(schema: &str) -> Result<Document, SchemaLoadError> {
    let borrowed_schema = graphql_parser::schema::parse_schema::<String>(&schema)?;
    Ok(schema_into_static(borrowed_schema))
}

fn schema_into_static<'a>(
    doc: graphql_parser::schema::Document<'a, String>,
) -> graphql_parser::schema::Document<'static, String> {
    // A workaround until https://github.com/graphql-rust/graphql-parser/pull/33 is
    // merged upstream.

    // A document that uses Strings for it's data should be safe to cast to 'static lifetime
    // as there's nothing inside it to reference 'a anyway
    unsafe { std::mem::transmute::<_, Document>(doc) }
}

#[derive(Debug, PartialEq, Clone)]
pub enum SchemaLoadError {
    IoError(String),
    ParseError(String),
    FileNotFound(String),
}

impl SchemaLoadError {
    pub fn into_syn_error(self, schema_span: proc_macro2::Span) -> syn::Error {
        let message = match self {
            SchemaLoadError::IoError(e) => format!("Could not load schema file: {}", e),
            SchemaLoadError::ParseError(e) => format!("Could not parse schema file: {}", e),
            SchemaLoadError::FileNotFound(e) => format!("Could not find file: {}", e),
        };

        syn::Error::new(schema_span, message)
    }
}

impl From<graphql_parser::schema::ParseError> for SchemaLoadError {
    fn from(e: graphql_parser::schema::ParseError) -> SchemaLoadError {
        SchemaLoadError::ParseError(e.to_string())
    }
}

impl From<std::io::Error> for SchemaLoadError {
    fn from(e: std::io::Error) -> SchemaLoadError {
        SchemaLoadError::IoError(e.to_string())
    }
}

/// Extension trait for the schema Field type
pub trait FieldExt {
    fn required_arguments(&self) -> Vec<InputValue>;
    fn optional_arguments(&self) -> Vec<InputValue>;
}

impl FieldExt for Field {
    fn required_arguments(&self) -> Vec<InputValue> {
        self.arguments
            .iter()
            .filter(|arg| {
                // Note: We're passing an empty TypeIndex in here, but that's OK as
                // we only want to know if things are required
                FieldArgument::from_input_value(arg, &TypeIndex::empty()).is_required()
            })
            .cloned()
            .collect()
    }

    fn optional_arguments(&self) -> Vec<InputValue> {
        self.arguments
            .iter()
            .filter(|arg| {
                // Note: We're passing an empty TypeIndex in here, but that's OK as
                // we only want to know if things are required
                !FieldArgument::from_input_value(arg, &TypeIndex::empty()).is_required()
            })
            .cloned()
            .collect()
    }
}

/// Extension trait for the schema Type type
pub trait TypeExt {
    fn to_graphql_string(&self) -> String;
    fn inner_name(&self) -> &str;
    fn is_required(&self) -> bool;
}

impl TypeExt for Type {
    fn to_graphql_string(&self) -> String {
        match self {
            Type::NamedType(name) => name.clone(),
            Type::ListType(inner) => format!("[{}]", inner.to_graphql_string()),
            Type::NonNullType(inner) => format!("{}!", inner.to_graphql_string()),
        }
    }

    fn inner_name(&self) -> &str {
        match self {
            Type::NamedType(s) => s,
            Type::ListType(inner) => inner.inner_name(),
            Type::NonNullType(inner) => inner.inner_name(),
        }
    }

    fn is_required(&self) -> bool {
        matches!(self, Type::NonNullType(_))
    }
}

pub trait ScalarTypeExt {
    fn is_builtin(&self) -> bool;
}

impl ScalarTypeExt for ScalarType {
    fn is_builtin(&self) -> bool {
        matches!(self.name.as_ref(), "String" | "Int" | "Boolean" | "ID")
    }
}
