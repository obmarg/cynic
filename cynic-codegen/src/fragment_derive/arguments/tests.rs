use proc_macro2::Span;
use rstest::rstest;
use syn::parse_quote;

use crate::schema::{parse_schema, types::Type, Schema};

use super::{analyse::analyse, parsing::CynicArguments};

#[rstest]
#[case::scalars("someScalarParams", parse_quote! { anInt: 1, aFloat: 3, anId: "hello" })]
#[case::missing_nullable_scalars("someNullableScalarParams", parse_quote! { anInt: 1 })]
#[case::empty_nullable_args("someNullableScalarParams", parse_quote! {})]
#[case::a_list_of_strings("filteredBooks", parse_quote! { filters: { authors: ["Charles Stross", "Ann Leckie"] } })]
#[case::list_wrapping("filteredBooks", parse_quote! { filters: { authors: "Ann Leckie" } })]
#[case::an_enum("filteredBooks", parse_quote! { filters: { state: "PUBLISHED" } })]
#[case::variable_in_object("filteredBooks", parse_quote! { filters: { state: $aVariable } })]
#[case::top_level_variable("filteredBooks", parse_quote! { filters: $aVaraible })]
#[case::top_level_variable("filteredBooks", parse_quote! { filters: $aVaraible, optionalFilters: $anotherVar })]
#[case::boolean_scalar("someNullableScalarParams", parse_quote! { aBool: true })]
#[case::missing_parameter("filteredBooks", parse_quote! {})]
#[case::unknown_parameter("someNullableScalarParams", parse_quote! { unknown: "hello" })]
fn test_analyse(#[case] field: &str, #[case] literals: CynicArguments) {
    use quote::format_ident;

    let schema_doc = parse_schema(SCHEMA).unwrap();
    let schema = Schema::new(&schema_doc).validate().unwrap();
    let ty = schema.lookup::<Type>("Query").ok().unwrap();
    let field = &ty.object().unwrap().field(field).unwrap();

    let literals = literals.arguments.into_iter().collect::<Vec<_>>();

    insta::assert_debug_snapshot!(analyse(
        literals,
        field,
        Some(&format_ident!("MyArguments").into()),
        Span::call_site()
    )
    .map(|o| o.arguments))
}

#[test]
fn test_analyse_errors_without_argument_struct() {
    let schema_doc = parse_schema(SCHEMA).unwrap();
    let schema = Schema::new(&schema_doc).validate().unwrap();
    let ty = schema.lookup::<Type>("Query").ok().unwrap();
    let field = &ty.object().unwrap().field("filteredBooks").unwrap();

    let literals: CynicArguments =
        parse_quote! { filters: $aVaraible, optionalFilters: $anotherVar };
    let literals = literals.arguments.into_iter().collect::<Vec<_>>();

    insta::assert_debug_snapshot!(
        analyse(literals, field, None, Span::call_site()).map(|o| o.arguments)
    )
}

const SCHEMA: &str = r#"
input BookFilters {
  authors: [String!]
  state: BookState
}

enum BookState {
  PUBLISHED
  OUT_OF_PRINT
}

type Query {
  filteredBooks(filters: BookFilters!, optionalFilters: BookFilters): [String!]!
  someScalarParams(anInt: Int!, aFloat: Float!, anId: ID!): [String]
  someNullableScalarParams(anInt: Int, aFloat: Float, anId: ID, aBool: Boolean): [String]
}
"#;
