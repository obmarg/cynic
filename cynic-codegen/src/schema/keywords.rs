use std::{borrow::Cow, collections::HashSet};

use once_cell::sync::Lazy;

static RAW_KEYWORDS: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    // A list of keywords in rust that can be converted to raw identifiers
    // Taken from https://doc.rust-lang.org/reference/keywords.html
    let mut set = HashSet::new();

    // Strict Keywords 2015
    set.insert("as");
    set.insert("break");
    set.insert("const");
    set.insert("continue");
    set.insert("else");
    set.insert("enum");
    set.insert("false");
    set.insert("fn");
    set.insert("for");
    set.insert("if");
    set.insert("impl");
    set.insert("in");
    set.insert("let");
    set.insert("loop");
    set.insert("match");
    set.insert("mod");
    set.insert("move");
    set.insert("mut");
    set.insert("pub");
    set.insert("ref");
    set.insert("return");
    set.insert("static");
    set.insert("struct");
    set.insert("trait");
    set.insert("true");
    set.insert("type");
    set.insert("unsafe");
    set.insert("use");
    set.insert("where");
    set.insert("while");

    // Strict keywords 2018
    set.insert("async");
    set.insert("await");
    set.insert("dyn");

    // Reserved Keywords 2015
    set.insert("abstract");
    set.insert("become");
    set.insert("box");
    set.insert("do");
    set.insert("final");
    set.insert("macro");
    set.insert("override");
    set.insert("priv");
    set.insert("typeof");
    set.insert("unsized");
    set.insert("virtual");
    set.insert("yield");

    // Reserved Keywords 2018
    set.insert("try");

    set
});

static NON_RAW_KEYWORDS: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    // A list of keywords in rust that cannot be converted to raw identifiers
    // Taken from https://github.com/rust-lang/rust/blob/1.31.1/src/libsyntax_pos/symbol.rs#L456-L460
    let mut set = HashSet::new();

    set.insert("super");
    set.insert("self");
    set.insert("Self");
    set.insert("extern");
    set.insert("crate");

    set
});

pub fn transform_keywords(s: &str) -> Cow<'_, str> {
    let s_ref: &str = s;
    if NON_RAW_KEYWORDS.contains(s_ref) {
        format!("{}_", s).into()
    } else if RAW_KEYWORDS.contains(s_ref) {
        format!("r#{}", s).into()
    } else {
        s.into()
    }
}
