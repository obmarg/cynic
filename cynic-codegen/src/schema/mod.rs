mod keywords;
mod names;
mod type_index;

pub mod parser;
pub mod types;

mod input;
pub mod markers;

use std::convert::Infallible;
use std::marker::PhantomData;

pub use self::{
    input::SchemaInput,
    names::FieldName,
    parser::{load_schema, Document},
};
use self::{type_index::TypeIndex, types::SchemaRoots};

// TODO: Uncomment this
// pub use self::{types::*},

pub struct Unvalidated;
pub struct Validated;

pub struct Schema<'a, State> {
    type_index: Box<dyn type_index::TypeIndex + 'a>,
    phantom: PhantomData<State>,
}

impl<'a> Schema<'a, Unvalidated> {
    pub fn new(input: SchemaInput) -> Self {
        let type_index: Box<dyn TypeIndex> = match input {
            SchemaInput::Document(document) => {
                Box::new(type_index::SchemaBackedTypeIndex::for_schema(document))
            }
            #[cfg(feature = "rkyv")]
            SchemaInput::Archive(data) => Box::new(
                type_index::optimised::ArchiveBacked::from_checked_data(data),
            ),
        };

        Schema {
            phantom: PhantomData,
            type_index,
        }
    }
}

impl<'a> Schema<'a, Unvalidated> {
    pub fn validate(self) -> Result<Schema<'a, Validated>, SchemaError> {
        self.type_index.validate_all()?;

        Ok(Schema {
            // doc: self.doc,
            type_index: self.type_index,
            phantom: PhantomData,
        })
    }

    pub fn lookup<'b, Kind>(&'b self, name: &str) -> Result<Kind, SchemaError>
    where
        Kind: TryFrom<types::Type<'b>> + 'b,
        Kind::Error: Into<SchemaError>,
    {
        Kind::try_from(self.type_index.lookup_valid_type(name)?).map_err(Into::into)
        // TODO: Suggestion logic should probably be implemented here (or in type_index)
    }
}

impl<'a> Schema<'a, Validated> {
    pub fn root_types(&self) -> Result<SchemaRoots<'_>, SchemaError> {
        self.type_index.root_types()
    }

    pub fn iter(&self) -> impl Iterator<Item = types::Type<'_>> {
        // unsafe_iter is safe because we're in a validated schema
        self.type_index.unsafe_iter()
    }

    // Looks up a kind that we're not certain is in the validated schema.
    pub fn try_lookup<'b, Kind>(&'b self, name: &str) -> Result<Kind, SchemaError>
    where
        Kind: TryFrom<types::Type<'b>, Error = SchemaError>,
    {
        Kind::try_from(self.type_index.lookup_valid_type(name)?)
        // TODO: Suggestion logic should probably be implemented here (or in type_index)
    }

    pub fn lookup<'b, Kind>(&'b self, name: &str) -> Result<Kind, SchemaError>
    where
        Kind: TryFrom<types::Type<'b>>,
        Kind::Error: Into<SchemaError>,
    {
        // unsafe_lookup is safe because we're validated
        Kind::try_from(self.type_index.unsafe_lookup(name).ok_or_else(|| {
            SchemaError::CouldNotFindType {
                name: name.to_string(),
            }
        })?)
        .map_err(Into::into)
    }
}

#[derive(Debug)]
pub enum SchemaError {
    UnexpectedKind {
        name: String,
        expected: types::Kind,
        found: types::Kind,
    },
    InvalidTypeInSchema {
        name: String,
        details: String,
    },
    CouldNotFindType {
        name: String,
    },
}

impl std::fmt::Display for SchemaError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::UnexpectedKind {
                name,
                expected,
                found,
            } => write!(
                f,
                "The type `{}` is a {} in the GraphQL schema but is being used where we expect a {}",
                name, found, expected,
            ),
            Self::InvalidTypeInSchema { name, details } => {
                write!(f, "Invalid schema when validating `{}`: {}", name, details)
            }
            Self::CouldNotFindType { name } => {
                write!(f, "Could not find a type named `{}` in the schema", name)
            }
        }
    }
}

impl From<SchemaError> for crate::Errors {
    fn from(val: SchemaError) -> Self {
        syn::Error::new(proc_macro2::Span::call_site(), val).into()
    }
}

impl SchemaError {
    pub fn unexpected_kind(actual: types::Type<'_>, expected: types::Kind) -> Self {
        SchemaError::UnexpectedKind {
            name: actual.name().to_string(),
            expected,
            found: actual.kind(),
        }
    }
}

impl From<Infallible> for SchemaError {
    fn from(_: Infallible) -> Self {
        panic!("This can't happen")
    }
}
