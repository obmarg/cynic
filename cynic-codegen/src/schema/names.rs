use crate::idents::RenamableFieldIdent;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct FieldName<'a> {
    pub(super) graphql_name: &'a str,
}

impl<'a> FieldName<'a> {
    pub fn as_str(&self) -> &'a str {
        self.graphql_name
    }
}

impl<'a> PartialEq<proc_macro2::Ident> for FieldName<'a> {
    fn eq(&self, other: &proc_macro2::Ident) -> bool {
        other == self.graphql_name
    }
}

impl<'a> PartialEq<str> for FieldName<'a> {
    fn eq(&self, other: &str) -> bool {
        self.graphql_name == other
    }
}

impl<'a> PartialEq<RenamableFieldIdent> for FieldName<'a> {
    fn eq(&self, other: &RenamableFieldIdent) -> bool {
        self.graphql_name == other.graphql_name()
    }
}
