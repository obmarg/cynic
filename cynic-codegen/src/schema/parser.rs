// TODO: It would be good to make these use &'str since this is compile time
// and max speed is best.

// TODO: It would be good to make these use &'str since this is compile time
// and max speed is best.

// Alias all the graphql_parser schema types so we don't have to specify generic parameters
// everywhere
pub type Document = graphql_parser::schema::Document<'static, String>;
pub type Type = graphql_parser::schema::Type<'static, String>;
pub type Field = graphql_parser::schema::Field<'static, String>;
pub type Definition = graphql_parser::schema::Definition<'static, String>;
pub type TypeDefinition = graphql_parser::schema::TypeDefinition<'static, String>;
pub type ScalarType = graphql_parser::schema::ScalarType<'static, String>;
pub type InputValue = graphql_parser::schema::InputValue<'static, String>;

/// Loads a schema from a string
pub fn load_schema(sdl: &str) -> Result<Document, SchemaLoadError> {
    Ok(add_typenames(parse_schema(sdl)?))
}

fn add_typenames(mut schema: Document) -> Document {
    for definition in &mut schema.definitions {
        match definition {
            Definition::TypeDefinition(TypeDefinition::Object(def)) => {
                def.fields.push(typename_field());
            }
            Definition::TypeDefinition(TypeDefinition::Interface(def)) => {
                def.fields.push(typename_field());
            }
            _ => {}
        }
    }
    schema
}

pub(crate) fn parse_schema(schema: &str) -> Result<Document, SchemaLoadError> {
    Ok(graphql_parser::schema::parse_schema::<String>(schema)?.into_static())
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum SchemaLoadError {
    IoError(String),
    ParseError(String),
    FileNotFound(String),
    NamedSchemaNotFound(String),
    DefaultSchemaNotFound,
    UnknownOutDirWithNamedSchema(String),
    UnknownOutDirWithDefaultSchema,
}

impl SchemaLoadError {
    pub fn into_syn_error(self, schema_span: proc_macro2::Span) -> syn::Error {
        syn::Error::new(schema_span, self.to_string())
    }
}

impl std::fmt::Display for SchemaLoadError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SchemaLoadError::IoError(e) => write!(f, "Could not load schema file: {}", e),
            SchemaLoadError::ParseError(e) => write!(f, "Could not parse schema file: {}", e),
            SchemaLoadError::FileNotFound(e) => write!(f, "Could not find file: {}", e),
            SchemaLoadError::NamedSchemaNotFound(_) => write!(
                f,
                "Could not find a schema with this name.  Have you registered it in build.rs?  {SCHEMA_DOCUMENTATION_TEXT}",
            ),
            SchemaLoadError::DefaultSchemaNotFound => {
                write!(f, "This derive is trying to use the default schema but it doesn't look like you've registered a default.  Please provide the `schema` argument or set a default in your build.rs.  {SCHEMA_DOCUMENTATION_TEXT}")
            }
            SchemaLoadError::UnknownOutDirWithNamedSchema(name) => {
                write!(f, "You requested a schema named {name} but it doesn't look like you've registered any schemas.  {SCHEMA_DOCUMENTATION_TEXT}")
            }
            SchemaLoadError::UnknownOutDirWithDefaultSchema => {
                write!(
                    f,
                    "This derive is trying to use the default schema, but it doesn't look like you've registered any schemas.  {SCHEMA_DOCUMENTATION_TEXT}"
                )
            }
        }
    }
}

// TODO: I could put a link in this
const SCHEMA_DOCUMENTATION_TEXT: &str =
    "See the cynic documentation on regsistering schemas if you need help.";

impl From<graphql_parser::schema::ParseError> for SchemaLoadError {
    fn from(e: graphql_parser::schema::ParseError) -> SchemaLoadError {
        SchemaLoadError::ParseError(e.to_string())
    }
}

impl From<std::io::Error> for SchemaLoadError {
    fn from(e: std::io::Error) -> SchemaLoadError {
        SchemaLoadError::IoError(e.to_string())
    }
}

/// Extension trait for the schema Type type
pub trait TypeExt {
    fn inner_name(&self) -> &str;
}

impl TypeExt for Type {
    fn inner_name(&self) -> &str {
        match self {
            Type::NamedType(s) => s,
            Type::ListType(inner) => inner.inner_name(),
            Type::NonNullType(inner) => inner.inner_name(),
        }
    }
}

pub trait ScalarTypeExt {
    fn is_builtin(&self) -> bool;
}

impl ScalarTypeExt for ScalarType {
    fn is_builtin(&self) -> bool {
        matches!(self.name.as_ref(), "String" | "Int" | "Boolean" | "ID")
    }
}

fn typename_field() -> Field {
    Field {
        position: graphql_parser::Pos { line: 0, column: 0 },
        description: None,
        name: "__typename".into(),
        arguments: vec![],
        field_type: Type::NonNullType(Box::new(Type::NamedType("String".into()))),
        directives: vec![],
    }
}
