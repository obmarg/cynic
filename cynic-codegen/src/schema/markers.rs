use std::borrow::Borrow;

use quote::format_ident;

use crate::schema::types::*;

use crate::idents::to_snake_case;

use super::keywords::transform_keywords;

/// Ident for a type
#[derive(Clone, Copy, Debug)]
pub struct TypeMarkerIdent<'a> {
    graphql_name: &'a str,
}

/// Ident for a field of a type
#[derive(Clone, Copy, Debug)]
pub struct FieldMarkerIdent<'a> {
    graphql_name: &'a str,
}

/// A module that contains everything associated with a field.
#[derive(Clone, Copy, Debug)]
pub struct FieldMarkerModule<'a> {
    type_name: &'a str,
}

/// A module that contains everything associated with an argument to a field
#[derive(Clone, Copy, Debug)]
pub struct ArgumentMarkerModule<'a> {
    type_name: &'a str,
    field_name: &'a str,
}

/// Marker to the type of a field - handles options & vecs and whatever the inner
/// type is
#[derive(Clone)]
pub struct TypeRefMarker<'a, T> {
    type_ref: &'a TypeRef<'a, T>,
}

impl<T> TypeRefMarker<'_, T> {
    pub fn to_path(&self, path_to_markers: &syn::Path) -> syn::Path {
        use syn::parse_quote;

        match &self.type_ref {
            TypeRef::Named(name, _, _) => {
                TypeMarkerIdent { graphql_name: name }.to_path(path_to_markers)
            }
            TypeRef::List(inner) => {
                let inner_path = TypeRefMarker {
                    type_ref: inner.as_ref(),
                }
                .to_path(path_to_markers);

                parse_quote! {
                    Vec<#inner_path>
                }
            }
            TypeRef::Nullable(inner) => {
                let inner_path = TypeRefMarker {
                    type_ref: inner.as_ref(),
                }
                .to_path(path_to_markers);

                parse_quote! {
                    Option<#inner_path>
                }
            }
        }
    }
}

impl<'a> TypeMarkerIdent<'a> {
    pub fn with_graphql_name(graphql_name: &'a str) -> Self {
        TypeMarkerIdent { graphql_name }
    }

    pub fn to_path(self, path_to_markers: &syn::Path) -> syn::Path {
        let mut path = path_to_markers.clone();
        path.push(proc_macro2::Ident::from(self));
        path
    }
}

impl From<TypeMarkerIdent<'_>> for proc_macro2::Ident {
    fn from(val: TypeMarkerIdent<'_>) -> Self {
        format_ident!("{}", transform_keywords(val.graphql_name))
    }
}

impl<'a> FieldMarkerModule<'a> {
    pub fn ident(&self) -> proc_macro2::Ident {
        format_ident!("{}_fields", to_snake_case(self.type_name))
    }

    pub fn to_path(self, path_to_markers: &syn::Path) -> syn::Path {
        let mut path = path_to_markers.clone();
        path.push(self.ident());
        path
    }
}

impl<'a> FieldMarkerIdent<'a> {
    pub fn to_path(self, path_to_markers: &syn::Path) -> syn::Path {
        let mut path = path_to_markers.clone();
        path.push(proc_macro2::Ident::from(self));
        path
    }
}

impl From<FieldMarkerIdent<'_>> for proc_macro2::Ident {
    fn from(val: FieldMarkerIdent<'_>) -> Self {
        if val.graphql_name == "_" {
            return format_ident!("_Underscore");
        }

        format_ident!("{}", transform_keywords(val.graphql_name))
    }
}

impl ArgumentMarkerModule<'_> {
    pub fn ident(&self) -> proc_macro2::Ident {
        format_ident!("{}_arguments", to_snake_case(self.field_name))
    }

    pub fn to_path(self, path_to_markers: &syn::Path) -> syn::Path {
        let mut path = path_to_markers.clone();
        path.push(
            FieldMarkerModule {
                type_name: self.type_name,
            }
            .ident(),
        );
        path.push(self.ident());
        path
    }
}

macro_rules! marker_ident_for_named {
    () => {};
    ($kind:ident) => {
        impl<'a> $kind<'a> {
            pub fn marker_ident(&self) -> TypeMarkerIdent<'a> {
                TypeMarkerIdent {
                    graphql_name: self.name
                }
            }
        }
    };
    ($kind:ident, $($rest:ident),+) => {
        marker_ident_for_named!($kind);
        marker_ident_for_named!($($rest),*);
    };
}

marker_ident_for_named!(
    ObjectType,
    EnumType,
    InterfaceType,
    UnionType,
    InputObjectType,
    ScalarType
);

impl<'a> Field<'a> {
    pub fn marker_ident(&self) -> FieldMarkerIdent<'a> {
        FieldMarkerIdent {
            graphql_name: self.name.as_str(),
        }
    }
}

impl<'a> InputValue<'a> {
    pub fn marker_ident(&self) -> FieldMarkerIdent<'a> {
        FieldMarkerIdent {
            graphql_name: self.name.as_str(),
        }
    }
}

impl<'a> ObjectType<'a> {
    pub fn field_module(&self) -> FieldMarkerModule<'a> {
        FieldMarkerModule {
            type_name: self.name,
        }
    }
}

impl<'a> InterfaceType<'a> {
    pub fn field_module(&self) -> FieldMarkerModule<'a> {
        FieldMarkerModule {
            type_name: self.name,
        }
    }
}

impl<'a> InputObjectType<'a> {
    pub fn field_module(&self) -> FieldMarkerModule<'a> {
        FieldMarkerModule {
            type_name: self.name,
        }
    }
}

impl<'a> Field<'a> {
    pub fn argument_module(&self) -> ArgumentMarkerModule<'a> {
        ArgumentMarkerModule {
            type_name: self.parent_type_name,
            field_name: self.name.graphql_name,
        }
    }
}

impl<'a> ObjectRef<'a> {
    pub fn marker_ident(&self) -> TypeMarkerIdent<'a> {
        TypeMarkerIdent {
            graphql_name: self.0,
        }
    }
}

impl<'a> InterfaceRef<'a> {
    pub fn marker_ident(&self) -> TypeMarkerIdent<'a> {
        TypeMarkerIdent {
            graphql_name: self.0,
        }
    }
}

impl<'a, T> TypeRef<'a, T> {
    pub fn marker_type(&'a self) -> TypeRefMarker<'a, T> {
        TypeRefMarker { type_ref: self }
    }
}

trait PathExt {
    fn push(&mut self, ident: impl Borrow<proc_macro2::Ident>);
}

impl PathExt for syn::Path {
    fn push(&mut self, ident: impl Borrow<proc_macro2::Ident>) {
        self.segments.push(ident.borrow().clone().into())
    }
}
