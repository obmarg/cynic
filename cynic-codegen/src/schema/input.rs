use crate::schema::load_schema;

use super::parser::SchemaLoadError;

pub enum SchemaInput {
    Document(cynic_parser::TypeSystemDocument),
    #[cfg(feature = "rkyv")]
    Archive(Vec<u8>),
}

// Public API
impl SchemaInput {
    /// Parses a SchemaInput from a filename passed to a macro
    pub(crate) fn default() -> Result<SchemaInput, SchemaLoadError> {
        Self::from_schema_name("default").map_err(|error| match error {
            SchemaLoadError::NamedSchemaNotFound(_) => SchemaLoadError::DefaultSchemaNotFound,
            SchemaLoadError::UnknownOutDirWithNamedSchema(_) => {
                SchemaLoadError::UnknownOutDirWithDefaultSchema
            }
            _ => error,
        })
    }

    /// Parses a SchemaInput from a name passed to a macro
    pub(crate) fn from_schema_name(name: &str) -> Result<SchemaInput, SchemaLoadError> {
        let out_dir = std::env::var("OUT_DIR")
            .map_err(|_| SchemaLoadError::UnknownOutDirWithNamedSchema(name.to_string()))?;

        let mut path = std::path::PathBuf::from(out_dir);
        path.push("cynic-schemas");
        #[cfg(feature = "rkyv")]
        let extension = "rkyv";
        #[cfg(not(feature = "rkyv"))]
        let extension = "graphql";
        path.push(format!("{name}.{extension}"));
        if !path.exists() {
            return Err(SchemaLoadError::NamedSchemaNotFound(name.to_string()));
        }

        #[cfg(feature = "rkyv")]
        {
            Self::from_rkyv_bytes(std::fs::read(path)?)
        }
        #[cfg(not(feature = "rkyv"))]
        {
            let string = std::fs::read_to_string(path)?;
            Self::from_sdl(&string)
        }
    }

    /// Parses a SchemaInput from a filename passed to a macro
    pub(crate) fn from_schema_path(
        path: impl AsRef<std::path::Path>,
    ) -> Result<SchemaInput, SchemaLoadError> {
        let path = path.as_ref();
        if let Some(ast) = document_from_path(path)? {
            return Ok(SchemaInput::Document(ast));
        }
        return Err(SchemaLoadError::FileNotFound(
            path.to_string_lossy().to_string(),
        ));
    }

    pub fn from_sdl(sdl: &str) -> Result<SchemaInput, SchemaLoadError> {
        load_schema(sdl).map(SchemaInput::Document)
    }

    #[cfg(feature = "rkyv")]
    pub fn from_rkyv_bytes(bytes: Vec<u8>) -> Result<SchemaInput, SchemaLoadError> {
        use crate::schema::type_index::optimised::OptimisedTypes;
        // Typecheck here so we can be sure it's safe later on
        rkyv::check_archived_root::<OptimisedTypes<'_>>(&bytes)
            .map_err(|e| SchemaLoadError::ParseError(e.to_string()))?;

        Ok(SchemaInput::Archive(bytes))
    }
}

fn document_from_path(
    filename: impl AsRef<std::path::Path>,
) -> Result<Option<cynic_parser::TypeSystemDocument>, SchemaLoadError> {
    use std::path::PathBuf;
    let mut pathbuf = PathBuf::new();

    if let Ok(manifest_dir) = std::env::var("CARGO_MANIFEST_DIR") {
        pathbuf.push(manifest_dir);
    } else {
        pathbuf.push(std::env::current_dir()?);
    }
    pathbuf.push(filename);

    if pathbuf.exists() {
        load_schema(&std::fs::read_to_string(pathbuf)?).map(Some)
    } else {
        Ok(None)
    }
}
