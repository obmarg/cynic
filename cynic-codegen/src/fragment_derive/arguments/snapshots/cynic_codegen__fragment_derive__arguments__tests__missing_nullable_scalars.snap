---
source: cynic-codegen/src/fragment_derive/arguments/tests.rs
expression: "analyse(&schema, literals, field, Some(&format_ident!(\"MyArguments\").into()),\n        Span::call_site()).map(|o| o.arguments)"
---
Ok(
    [
        Field {
            schema_field: InputValue {
                name: FieldName {
                    graphql_name: "anInt",
                },
                value_type: NullableType(
                    NamedInputType(
                        "Int",
                    ),
                ),
                has_default: false,
            },
            value: Some(
                Literal(
                    Lit::Int {
                        token: 1,
                    },
                ),
            ),
        },
    ],
)
